### PR TITLE
Introduce anomaly filters

### PR DESCRIPTION
**Pull request by MaxHalford**
This work addresses #931. But as usual this was a bit of rabbit hole, so I did other (necessary) changes.

The main addition is the `anomaly.base.AnomalyFilter` base class. The idea is that we want to detect anomalous data, so that we can "skip it" to prevent a downstream model from learning on anomalous data. In short, you can now do this:

```py
model = (
    anomaly.ThresholdFilter(
        anomaly.GaussianScorer(
            window_size=24 * 7,
            grace_period=30
        ),
        threshold=0.995
    ) |
    time_series.HoltWinters(
        alpha=0.3,
        beta=0.1,
        multiplicative=False
    )
)
```

The intelligence happens in the `Pipeline.learn_one` method:

```py
if isinstance(t, anomaly.base.AnomalyFilter):
    if t._supervised:
        t.learn_one(x, y)
        score = t.score_one(x, y)
    else:
        t.learn_one(x)
        score = t.score_one(x)
    # Skip the next parts of the pipeline if the score is classified as anomalous
    if t.classify(score):
        is_anomaly = True
        break
    continue
```

In @dberardo-com's case, it's a time series problem. His company kindly shared with us a new dataset called `WaterFlow`. It's a univariate time series. Instead of awkwardly making our existing unsupervised anomaly detectors fit the bill, I decided it would make more sense to introduce `anomaly.base.SupervisedAnomalyDetector`.

I added a supervised anomaly detection method based on @pranab's work in #829 on using a Gaussian distribution. I'm sorry your PR was refused at the time @pranab! Our design was not mature enough yet. I feel much more confident now that we this notion of supervised anomaly detectors. It's the perfect way to add univariate anomaly detection methods to River. @Fengrui-Liu you might want to take a close look at this too :)

I added an example in the docstring of `anomaly.ConstantThresholder` which trains a Holt-Winters model on the water flow dataset. Here's some before and after charts:

Screenshot 2022-05-22 at 17 20 28

Screenshot 2022-05-22 at 17 20 11

Of course, it looks like the first chart is better. But it isn't. The lower spikes in red happen after the green spikes. That's because the green spikes are corrupting the model. The model in the second chart is more robust.

A nice consequence of introducing `anomaly.base.AnomalyFilter` is that it kind of deprecates the need for `anomaly.ConstantThresholder` and `anomaly.QuantileThresholder`. I removed them because else the `anomaly` module was starting to look a bit crowded.

I went through 2 or 3 designs before settling on this one. I'm pretty happy with the basic building blocks we have now. We're much more mature with regards to anomaly detection, as guarding models against anomalous data is usecase that comes up quite often. We can iterate and add additional methods and examples with these building blocks in place.

### CODE REVIEW

**Comment by MaxHalford** [comment]
@smastelini I'll let you review the PR! Don't feel the need to go deep into it :)

**Comment by MaxHalford** [comment]
The failing tests are the usual suspects. They'll get fixed once #927 is merged :)

**Comment by Fengrui-Liu** [comment]
Interesting and important question! This is very helpful to me.

I still wonder how it will behave when anomalies meet concept drift, here is an example

image

Assume that the anomaly filter ignores updating the data after 100 (a concept drift occurs), will it act as predict1 ? Do we need to handle this complicated situation?

**Comment by MaxHalford** [comment]
Good question! My take is that concept drift should be handled differently. A concept drift is not an anomaly. The model simply has to deal with it and adapt. That is in fact the whole premise and benefit of online machine learning.

In practice, we should expect the anomaly detector to treat the first samples after the concept drift as anomalous, and will prevent the downstream model from learning. Then it will start treating the data as normal, and re-allow the model to learn. In your example, that would result in a bit of red followed by green.

**Comment by Fengrui-Liu** [comment]
> Then it will start treating the data as normal, and re-allow the model to learn. In your example, that would result in a bit of red followed by green.

So, just the filters help us to ignore the anomalies, and the anomaly detector can still learn/update from those anomalies? If not, the detector should regard the concept drift data as anomalies persistently, and how does the detector retreat the concept drift data as normal without learning from them?

I'm still a little bit confused, and going to read the code in detail after this PR 🍀 .

**Comment by dberardo-com** [comment]
thanks for engineering this useful feature in such low time @MaxHalford glad to know that the dataset helped.

Here quite a bit of comments on the PR from myself. As i am still new to River codebase, please take my comments with a pinch of salt.

### comments on the core changes

I went through the PR myself too, and i like the fact that the change in the pipeline.py file feels very "natural" so that should be a good sign :D  

I forgot to mention that expliclty in the feature request issue, but yes the anomalies detector should also learn from the incoming data, why not ?

@Fengrui-Liu  the anomalies detector is also trained as part of the same pipeline: https://github.com/online-ml/river/pull/936/files#diff-f5f1adf689cb04893a0cae749d3040721b16258ba1641e41cc7d74b93b508e83R488

https://github.com/online-ml/river/pull/936/files#diff-030a82ec8ba455c71397118e237a54593e4a870e8ce2cf06076f5d2f1d2906a2R142 --> in this class however, anomalies are excluded from learning, so how could the detector actually learn that drift have occured, in the example above this means that all green samples are excluded from learning after timestep 100 and thus always considered anomalies, isnt it?

### comments on anomalies detector

I am definitely not an expert, but it is not fully clear what it is meant with "supervised detector". When i think about supervised learning i think about a dataset having features and labels. In this case the labels should be whether the data sample is an actual anomaly or not. I could not see this label injection in the Gaussian detector, so perhaps i am missing the "supervised nature" of it: https://github.com/online-ml/river/pull/936/files#diff-bf4876a642151b03bc28164477ec162492c0ad557280a32c99d71abcbcd81fa6R8

In the dataset i provided, the data was hourly sampled, but in reality data comes through at very unpredictable frequencies. Also, the dataset shows one gap by the end of the 2 months period. In both cases, it is not enough to assume a fixed number of points (here 24*7) as a window_size for the GaussianScorer: https://github.com/online-ml/river/pull/936/files#diff-42998e0ee68c89b131fd52621cdc6c57684b3288936ff8d0a889a4a52dc1e6baR34
Question here would be --> is it possible for the Scorer to look into actual timestamps rather than using counter on values? I am not sure whether River has this concept but in Spark Structured Streaming there is also a concept called Watermarking, which defines a sort of grace period for late arriving observations. Also for in the latest case it is not enough to count samples, but to actually look at their timestamp.

### comments on the evaluator

Are anomalies removed also when evaluating predictions? this would have a big impact on the final scoring

### comments on the deprecation

I dont know River's release / versioning policy but i am bit concerned with the removal of the ConstantThresholder class. Reoving such an elemntar building block could be quite of a breaking change and might scare some adopters of River when assessing "long-term support and stability" of this library. I would rather go for a deprecation warning when importing the class and then casting that class in the backend as an instance of ThresholdFilter instead. That would make the changes backward compatibles.

P.S. in this file it is not clear wheather ConstantThresholder is added or removed: https://github.com/online-ml/river/pull/936/files#diff-fe8ea9edcd053664db896bac1abd739e8c9d3ab234e94a4113fa97c83ac1a333R9

P.S.2 i might have screwed up the hyperparameters completely, but when using SVM for anomalies detection i actually got high negative scores sometimes. In this sense this phrase is somewhat misleading, because assumes postivie scoring values (https://github.com/online-ml/river/pull/936/files#diff-bf4876a642151b03bc28164477ec162492c0ad557280a32c99d71abcbcd81fa6R5) --> i am being very peaky here, so ignore this in case

**Comment by MaxHalford** [comment]
> So, just the filters help us to ignore the anomalies, and the anomaly detector can still learn/update from those anomalies? If not, the detector should regard the concept drift data as anomalies persistently, and how does the detector retreat the concept drift data as normal without learning from them?

Very good point. Both of you are right. We should probably make this "guard rail" behavior optional. Ok with both of you?

> In this case the labels should be whether the data sample is an actual anomaly or not

No, in the case of supervised anomaly detection, the label is simply the label of the underlying dataset. The "data sample is an actual anomaly or not" is not in fact known.

> A. In the dataset i provided, the data was hourly sampled, but in reality data comes through at very unpredictable frequencies.

Good catch! It's definitely possible to use a time rolling window, but that will require quite a bit more refactoring. It's [included](https://github.com/online-ml/river/blob/main/river/proba/time_rolling.py) in the new release.

> B. Also, the dataset shows one gap by the end of the 2 months period.

I actually removed that last gap because I thought it would be confusing for new users.

> do you want me to provide you the real, unsampled waterflow dataset, so you could see the impact on variable frequency on the prediction results? --> is this a concern of the data streaming backbone (e.g. kafka) to sample the data or should we use online aggregators in river in front of the pipeline?

Not sure I have the full answer, but time series models usually assume a uniform sampling scheme. The model would likely perform better if the data was sampled hourly.

> this is probably matter for another issue?

I would say so.

> Are anomalies removed also when evaluating predictions? this would have a big impact on the final scoring

They are not, no. I thought about that, but I couldn't think about an easy way to implement that. I guess that's also work for another PR. You might argue that some users would want to keep anomalies.

> I dont know River's release / versioning policy but i am bit concerned with the removal of the ConstantThresholder class. Reoving such an elemntar building block could be quite of a breaking change and might scare some adopters of River when assessing "long-term support and stability" of this library. I would rather go for a deprecation warning when importing the class and then casting that class in the backend as an instance of ThresholdFilter instead. That would make the changes backward compatibles.

The behavior of `ConstantThresholder` is not lost, and is part of the `ConstantFilter` class. I agree that it's not the same API.

I agree that removing features between versions is not ideal. But for the moment it's the way I prefer to proceed. River is still a small library. I rather be nimble and have less work to do while we are a small library. I already have very little time to work on River, and I just haven't taken the time to put in place a deprecation system. Maybe I should. For the while, users simply have to make the necessary changes before bumping a version. It's more work for users, and less work for us developers. It's a trade-off. Also, until now I've never had a complaint, which is not to say this way of proceeding hasn't bothered anyone.

> P.S.2 i might have screwed up the hyperparameters completely, but when using SVM for anomalies detection i actually got high negative scores sometimes. In this sense this phrase is somewhat misleading, because assumes postivie scoring values (https://github.com/online-ml/river/pull/936/files#diff-bf4876a642151b03bc28164477ec162492c0ad557280a32c99d71abcbcd81fa6R5) --> i am being very peaky here, so ignore this in case

I don't think "low" or "high" indicates whether the score is positive or not. As said afterwards, the range of the score depends on the method. Do you have a better formulation to suggest?

**Comment by dberardo-com** [comment]
> Very good point. Both of you are right. We should probably make this "guard rail" behavior optional. Ok with both of you?

Guard rail --> you mean the self-learning of the anomalies detector. I would keep this on by default, but if we want to allow for it to got "off" at one point, then a flag is not enough, since the anomalies detector needs to have gathered some least amount of points before becoming useful, so the user should define a "autoUpdatePolicy: 'ignore-drifts" / 'every week' / 'others' " otherwise how to make sure to have a usable anomalies detector backing the pipeline ? --> i hope i understood the problem here :D 

> No, in the case of supervised anomaly detection, the label is simply the label of the underlying dataset. The "data sample is an actual anomaly or not" is not in fact known.

--> this is my "non-expert" definition of unsupervised learning :D if one does not know the labels ... right? 

> Good catch! It's definitely possible to use a time rolling window, but that will require quite a bit more refactoring. It's [included](https://github.com/online-ml/river/blob/main/river/proba/time_rolling.py) in the new release.

--> i had to implement this myself in some notebooks and face the fact the the aggregator's state was increasing with time, since i was bucketing samples based on hours (which might be wrong). I had to impose a simple policy such that periodically removed old buckets from the model.state dictionary of the aggregator. This can be considered a rudimental implementation of a Watermakr in my opinion.

> They are not, no. I thought about that, but I couldn't think about an easy way to implement that. I guess that's also work for another PR. You might argue that some users would want to keep anomalies.

i thought this would be a tough one, but it needs to be kept in mind for future work, because otherwise a model selection algorithm would always go for the "overfitting" anomaly-detector-less version of the pipeline, wouldnt it?

 > Also, until now I've never had a complaint, which is not to say this way of proceeding hasn't bothered anyone.

--> ok, but then in theory this would imply a "major" release update according to semantic versioning. I completely understand your points though

> I don't think "low" or "high" indicates whether the score is positive or not. As said afterwards, the range of the score depends on the method. Do you have a better formulation to suggest?

totally fine with that. I understand that is easy enough to digest for new users, thus to be preferred. Probably one could use the word "magnitude" as in high magnitude and low magnitude ... but that sounds weird, so forget it for now. IMO

**Comment by Fengrui-Liu** [comment]
cannot agree more 👍

**Comment by dberardo-com** [comment]
> I think it's simpler than that, at least at first. The question is very simple: should the anomaly filter allow the anomaly detector to be updated with every anomaly score. At the moment the answer is no. But it could be yes. A simple boolean flag will allow that.

> Indeed, we have to allow the anomaly detector to be updated. e.g. the flag I suggested above. What do you think?

Ok, now it all makes sense :D  --> i would agree that the default should be to keep the learning on, although i am biased by my own use case here. However, i still have to reflect a bit on what would happen when the learning is off ... as i cannot imagine on the basis of what the detector would score incoming data. But i will run some tests with other datasets to figure that out.

Also, i am still new to the concept of drifting, and thus i dont know if it would still make sense to monitor drift outside the pipeline. But i will try to assess that during tests too.

Question here: should this "adaptive anomalies detector flag" be an initialization constant/argument or can it be switched on/off programmatically as required? The least gives us more flexibility i guess

---

> A dataset is a sequence of (x, y) pairs. x are the features and y are the labels. The anomaly detectors we have right now are unsupervised, and so only learn from x. I'm proposing to introduce supervised anomaly detectors, which would also take into account y. But that doesn't mean y are the labels indicating whether or not y is an anomaly. They're just labels, such as numbers or classes. Is that clearer?

I guess i am being "blind" here and consider only univariate datasets in my mental process. In case of univariate i cannot understand the difference between x and y. But i guess that difference does exist for richer feature vectors x. If this is the case then i understand the matter now, as i was not aware that other detectors do not take labels into account.

---

> Of course, you're right. Let's tackle that in a separate PR.

ok so all in all after our discussion i see 3 actions here:

1. on this PR --> include programmatic swtch / function argument for A-detector learning
2. on another PR/issue --> figure out how to include anomalies when evaluating forecaster
3. on another issue --> figure out when and how to aggregate data samples and whether to consider Watermarking, or have both count based and timestamp based window_size in GaussianScorer

am i right?

Do you want me to kick-off point number 3?

**Comment by MaxHalford** [comment]
> Question here: should this "adaptive anomalies detector flag" be an initialization constant/argument or can it be switched on/off programmatically as required? The least gives us more flexibility i guess

It's a parameter, so it's specified when the class is instantiated. But all parameters in River are stored as attributes, so you can easily change afterwards.

> In case of univariate i cannot understand the difference between x and y

Indeed, the thing is that not all problems are time series problem where `x` is virtually empty. In many supervised, non time series, problems, detecting that `y` is an anomaly can be very useful. In other words, the use case you brought to us will reap benefits in other use cases :)

> on this PR --> include programmatic swtch / function argument for A-detector learning

Yep, I'll do that.

> on another PR/issue --> figure out how to include anomalies when evaluating forecaster

I think you mean exclude, but yes.

> on another issue --> figure out when and how to aggregate data samples and whether to consider Watermarking, or have both count based and timestamp based window_size in GaussianScorer

Yep! Sounds just about right.

> Do you want me to kick-off point number 3?

Sure! It sounds complicated though, so don't feel the obligation to work on it right now.

**Comment by dberardo-com** [comment]
ok for all other the points 

> Sure! It sounds complicated though, so don't feel the obligation to work on it right now.

would just be a rough description of requirements into a feature request. more details would follow in the description.

Looking forward for the next release of river in pip ;)

**Comment by MaxHalford** [comment]
> Looking forward for the next release of river in pip ;)

Me too! It'll likely happen once #927 is merged :)

File: docs/releases/unreleased.md [hunk 1]
```diff
@@ -2,6 +2,18 @@
 
 - Moved all metrics in `metrics.cluster` except `metrics.Silhouette` to [river-extra](https://github.com/online-ml/river-extra).
 
+## anomaly
+
+- There is now a `anomaly.base.SupervisedAnomalyDetector` base class for supervised anomaly detection.
+- Added `anomaly.GaussianScorer`, which is the first supervised anomaly detector.
+- There is now a `anomaly.base.AnomalyFilter` base class for anomaly filtering methods. These allow to classify anomaly scores. They can also prevent models from learning on anomalous data, for instance by putting them as an initial step of a pipeline.
+- Added `anomaly.ConstantThresholder`, which is the first anomaly filter.
```
Comment (smastelini) [inline docs/releases/unreleased.md:10]: Filter?

**Review by smastelini — COMMENTED** [review]
Great work @MaxHalford ! I left two comments for future discussion :)

File: river/anomaly/gaussian.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+from river import anomaly, proba
+
+
+class GaussianScorer(anomaly.base.SupervisedAnomalyDetector):
```
[the remark this answers is not in the document]
Comment (MaxHalford) [reply river/anomaly/gaussian.py:4]: > Should we think about a way to differentiate univariate and multivariate anomaly detectors?

For the moment I would say no! I would say that univariate detectors can be implemented as supervised anomaly detectors. Having them as unsupervised estimators is tricky because `x` would have to be a single value rather than a `dict`.

**Comment by dberardo-com** [comment]
i was researching on the matter of "supervised anomalies detection" and i found this paper: https://arxiv.org/ftp/arxiv/papers/1401/1401.6424.pdf

however there i can see that they also mean "supervised" as in "using labeled anmalies samples". so some confusion arises. Would you mind sharing a source with the definition that River would use for supervised anomalies detection? @MaxHalford

**Comment by MaxHalford** [comment]
> however there i can see that they also mean "supervised" as in "using labeled anmalies samples".

That just sounds like binary classification.

> Would you mind sharing a source with the definition that River would use for supervised anomalies detection? @MaxHalford

I don't really have any. The only thing you have to understand is that almost all anomaly detection algorithms operate on the features of a dataset. So in that sense they're unsupervised. But I also want us to have anomaly detection algorithms which can detect anomalies in labels. In particular, this includes many univariate algorithms which can be done online.

This isn't set in stone. If we find a better way to frame things, then let's do it. I'm just trying to think of a way of doing which users might enjoy.
